### PR TITLE
Feature/achievement badge/vm

### PR DESCRIPTION
**Pull request by vmmuthu31**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new API endpoint for updating user achievement badges.
	- Added functionality to retrieve and display achievement badges in user profiles.
	- Enhanced user profile structure to include achievement badges.
	- Implemented new GraphQL queries for detailed voter and delegation data retrieval.

- **Bug Fixes**
	- Improved error handling for badge evaluation and updates.

- **Documentation**
	- Added descriptions for new properties and functions related to achievement badges.

### CODE REVIEW

File: pages/api/v1/achivementbadges/update-badges.ts [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+		return;
+	}
+	const userId = user.user_id;
+	const userDocRef = firestore_db.collection('users').doc(userId.toString());
+	const newBadges = await evaluateBadges(username, network);
+	const existingBadges = user.achievement_badges || [];
+	const validExistingBadges = existingBadges.filter((existingBadge) => newBadges.some((newBadge) => newBadge.name === existingBadge.name));
+	const filteredNewBadges = newBadges.filter((newBadge) => !existingBadges.some((existingBadge) => existingBadge.name === newBadge.name));
+	const updatedBadges = [...validExistingBadges, ...filteredNewBadges];
+	const updateData: any = {
```
Comment (alphainfinitus) [inline pages/api/v1/achivementbadges/update-badges.ts:339]: please defer from using `any` when we know the types

File: pages/api/v1/achivementbadges/utils.ts
Comment (alphainfinitus) [inline pages/api/v1/achivementbadges/utils.ts:1]: this file will be treated as an api route

File: pages/user/[username].tsx [hunk 2]
```diff
@@ -71,6 +72,9 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
 		userId: userProfile?.data?.user_id
 	});
 
+	const achievement_badges = await updateUserBadges(username.toString(), network);
+	console.log('achievement_badges', achievement_badges);
```
Comment (alphainfinitus) [inline pages/user/[username].tsx:76]: lingering console log

File: pages/api/v1/achivementbadges/constant.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }`;
+
+export const GET_PROPOSAL_COUNT = `query ProposalCountQuery($startDate: DateTime!) {
+    proposalsConnection(where: { createdAt_gte: $startDate, type_in: [ReferendumV2, Referendum] }, orderBy: id_DESC) {
+        totalCount
+    }
+}`;
+
+export const badgeNames = ['Decentralised Voice', 'Fellow', 'Council Member', 'Active Voter', 'Whale', 'Steadfast Commentor', 'GM Voter', 'Popular Delegate'];
```
Comment (alphainfinitus) [inline pages/api/v1/achivementbadges/constant.ts:62]: badgenames is stored here as well as in `src/global/achievementbadges.ts`

File: src/global/achievementbadges.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	img: string;
+	name: BadgeName;
+	requirements: string;
+}
+
+export const badgeDetails: BadgeDetails[] = [
+	{
+		description: 'Awarded to Polkadot delegates who have significant influence.',
+		id: '1',
+		img: '',
```
Comment (alphainfinitus) [inline src/global/achievementbadges.ts:19]: no images have been added for prod

File: src/global/achievementbadges.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		name: BadgeName.DecentralisedVoice_polkodot,
+		requirements: 'Must be a delegate on the Polkadot network.'
+	},
+	{
+		description: 'Awarded to Kusama delegates who have significant influence.',
+		id: '2',
```
Comment (alphainfinitus) [inline src/global/achievementbadges.ts:25]: i don't think ids are required here

File: src/auth/types.ts [hunk 6]
```diff
@@ -261,3 +262,37 @@ export interface SubscanAPIResponseType {
 	url?: number;
 	body?: any;
 }
+
+export enum BadgeName {
```
Comment (alphainfinitus) [inline src/auth/types.ts:266]: enum does not follow standard as other enums

File: pages/api/v1/achivementbadges/utils.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+import { MessageType } from '~src/auth/types';
+import { IDelegationStats } from '../delegations/get-delegation-stats';
+import nextApiClientFetch from '~src/util/nextApiClientFetch';
+import BN from 'bn.js';
+import { ApiPromise, WsProvider } from '@polkadot/api';
+import { chainProperties } from '~src/global/networkConstants';
+
+export async function getDelegationStats(network: string): Promise<{ data: IDelegationStats }> {
+	const { data, error } = await nextApiClientFetch<IDelegationStats | MessageType>(`/api/v1/delegations/get-delegation-stats?network=${network}`);
```
Comment (alphainfinitus) [inline pages/api/v1/achivementbadges/utils.ts:13]: `nextApiClientFetch` is to be only used on the Client side

**Review by alphainfinitus — CHANGES_REQUESTED** [review]

File: src/types.ts [hunk 8]
```diff
@@ -671,6 +671,15 @@ export enum EActivityFilter {
 	REACTS = 'REACTED',
 	MENTIONS = 'MENTIONED'
 }
+
+export enum EProfileBadges {
```
Comment (alphainfinitus) [inline src/types.ts:675]: this enum and  BadgeName enum seem redundant

**Review by alphainfinitus — CHANGES_REQUESTED** [review]

File: src/components/UserProfile/ProfileBadges.tsx [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	className?: string;
+	theme?: string;
+	badges?: Badge[];
+	addressWithIdentity?: string;
+	userProfile: ProfileDetailsResponse;
+	selectedAddresses: string[];
+}
+
+const defaultLockedBadges = [
+	{ img: '/assets/badges/active_voter_locked.svg', name: BadgeName.ACTIVE_VOTER },
```
Comment (alphainfinitus) [inline src/components/UserProfile/ProfileBadges.tsx:26]: locked image should be stored with other badge data

File: pages/api/v1/achivementbadges/update-badges.ts [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+
+// Badge2: Check if the user qualifies for the Fellow badge based on rank
+async function checkFellow(user: ProfileDetailsResponse, network: string): Promise<boolean> {
+	if (!user?.addresses || user.addresses.length === 0 || !isValidNetwork(network)) {
+		console.warn(`No addresses found for user: ${user.username}`);
+		return false;
+	}
+
+	// Only run for collectives or westend-collectives networks
+	if (network !== 'collectives' && network !== 'westend-collectives') {
```
Comment (alphainfinitus) [inline pages/api/v1/achivementbadges/update-badges.ts:52]: fellow badge will only be evaluated if visited from fellowship network

**Review by alphainfinitus — CHANGES_REQUESTED** [review]

**Review by alphainfinitus — APPROVED** [review]

File: pages/api/v1/achivementbadges/update-badges.ts [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+	}
+	const profile: ProfileDetails = userDoc.data()?.profile || {};
+
+	const existingBadges = profile.achievement_badges || [];
+
+	const validBadges = existingBadges.filter((existingBadge: Badge) => newBadges.some((newBadge: Badge) => newBadge.name === existingBadge.name));
+
+	const updatedBadges = [...validBadges, ...newBadges];
+
+	const updateData = {
```
Comment (KanishkaRajputd) [inline pages/api/v1/achivementbadges/update-badges.ts:305]: {profile:{...userDoc?.data(),achievement_badges:updatedBadges }

**Review by KanishkaRajputd — CHANGES_REQUESTED** [review]

**Pull request merged**